design/kafka-upgrade/README.md

```diff
@@ -0,0 +1,175 @@
+# Supporting Kafka upgrades
+
+* Status: **Under discussion**
+* Discussion: [GitHub PR](https://github.com/strimzi/strimzi-kafka-operator/pull/623)
+
+## Motivation
+
+The Strimzi Cluster Operator (CO) will need to support upgrading the Kafka 
+version used on a cluster-by-cluster basis. 
+(Note this is a different thing to supporting Operator upgrades).
+
+## Scope
+
+The expectation is that a given version of the operator will support some 
+defined set of Kafka versions and the user can selector which should be 
+used for their cluster.
+When the user upgrages to a later version supported by the CO, the CO will upgrade the cluster in a zero-downtime manner.
+
+Support for downgrading to an earlier version is not a priority.
+
+Upgrading Zookeeper is out of scope. We will assume that the new version of Kafka can use the existing Zookeeper version.
```

Comment:
Today we have a `kafka-base` image which brings the Kafka and Zookeeper versions we are going to use. Does it mean that they can diverge? I mean, I'm using a `kafka-base` with newer Kafka and Zookeeper versions but using only the Kafka part and the an old `kafka-base` image used for Zookeeper?

Reply:
TBH, I put this in just because I didn't want to think about ZK upgrade yet. Having done a tiny bit of research, it looks like with ZK you would normally just need to do a rolling update with the new version, but that's not always true (e.g. https://zookeeper.apache.org/doc/r3.5.3-beta/zookeeperReconfig.html#ch_reconfig_upgrade but that's 3.4-> 3.5 and Kafka's still on 3.4).

If it is this simple, at least for current versions of ZK used by Kafka, I would be happy to do this too to avoid this kind of problem. Obviously transiently, during the upgrade, we end up running Kafka against a different version of ZK than it's really meant to, but that's unavoidable with the no-downtime constraint.